File: src/protocols/obsidian/frontmatter.ts
/**
 * Copyright 2026 Daniel Smith — Apache 2.0
 *
 * YAML frontmatter extraction for `.md` notes.
 *
 * Uses the `yaml` package when installed (full spec support), and falls
 * back to a line-based parser that handles the keys we actually read
 * (title, tags, cover, image, description, created, modified). The
 * fallback is intentionally minimal — if a vault's frontmatter uses
 * anchors or complex nested structures, install `yaml`.
 */

import { loadOptional } from '../shared';
import { OBSIDIAN_FRONTMATTER_FENCE, OBSIDIAN_OPTIONAL_YAML_PKG } from './constants';

export interface ParsedMarkdown {
  frontmatter: Record<string, unknown>;
  body: string;
}

interface YamlModule {
  parse: (src: string) => unknown;
}

/**
 * Split a markdown buffer into `{ frontmatter, body }`.
 * Notes without a leading `---` fence get `frontmatter: {}` and the full
 * text as the body.
 */
export const parseMarkdown = async (source: string): Promise<ParsedMarkdown> => {
  const fence = OBSIDIAN_FRONTMATTER_FENCE;
  if (!source.startsWith(`${fence}\n`) && !source.startsWith(`${fence}\r\n`)) {
    return { frontmatter: {}, body: source };
  }

  // Find the closing fence — must appear on its own line.
  const closeRegex = new RegExp(`\\r?\\n${escapeRegex(fence)}\\r?\\n`);
  const rest = source.slice(fence.length + 1);
  const closeMatch = rest.match(closeRegex);
  if (!closeMatch || closeMatch.index === undefined) {
    return { frontmatter: {}, body: source };
  }

  const yamlText = rest.slice(0, closeMatch.index);
  const body = rest.slice(closeMatch.index + closeMatch[0].length);

  const frontmatter = await parseYamlBlock(yamlText);
  return { frontmatter, body };
};

const parseYamlBlock = async (yamlText: string): Promise<Record<string, unknown>> => {
  const mod = await loadOptional<YamlModule>(OBSIDIAN_OPTIONAL_YAML_PKG);
  if (mod?.parse) {
    try {
      const parsed = mod.parse(yamlText);
      return parsed && typeof parsed === 'object' && !Array.isArray(parsed)
        ? (parsed as Record<string, unknown>)
        : {};
    } catch {
      // Bad YAML — fall through to the minimal parser so we still return *something*.
    }
  }
  return parseMinimalYaml(yamlText);
};

/**
 * Minimal YAML subset: top-level `key: value` pairs plus tag arrays.
 * Supports:
 *   - scalars (strings, numbers, bools)
 *   - quoted strings ("...", '...')
 *   - inline arrays (`tags: [a, b, c]`)
 *   - block arrays:
 *       tags:
 *         - a
 *         - b
 *
 * Anything more complex returns as the raw string — callers treat unknown
 * shapes defensively.
 */
const parseMinimalYaml = (text: string): Record<string, unknown> => {
  const out: Record<string, unknown> = {};
  const lines = text.split(/\r?\n/);
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];
    i++;
    if (!line || /^\s*#/.test(line)) continue;

    const kvMatch = line.match(/^([A-Za-z_][A-Za-z0-9_-]*)\s*:\s*(.*)$/);
    if (!kvMatch) continue;
    const key = kvMatch[1];
    const raw = kvMatch[2].trim();

    if (raw === '') {
      // Possible block array on subsequent indented lines
      const arr: unknown[] = [];
      while (i < lines.length) {
        const next = lines[i];
        const itemMatch = next.match(/^\s*-\s+(.*)$/);
        if (!itemMatch) break;
        arr.push(coerceScalar(itemMatch[1].trim()));
        i++;
      }
      out[key] = arr;
      continue;
    }

    if (raw.startsWith('[') && raw.endsWith(']')) {
      out[key] = raw
        .slice(1, -1)
        .split(',')
        .map((s) => coerceScalar(s.trim()))
        .filter((v) => v !== '');
      continue;
    }

    out[key] = coerceScalar(raw);
  }

  return out;
};

const coerceScalar = (raw: string): unknown => {
  if (raw === '') return '';
  if ((raw.startsWith('"') && raw.endsWith('"')) || (raw.startsWith("'") && raw.endsWith("'"))) {
    return raw.slice(1, -1);
  }
  if (raw === 'true') return true;
  if (raw === 'false') return false;
  if (raw === 'null' || raw === '~') return null;
  if (/^-?\d+$/.test(raw)) return Number(raw);
  if (/^-?\d+\.\d+$/.test(raw)) return Number(raw);
  return raw;
};

const escapeRegex = (s: string): string => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

File: scripts/lib/vault_from_md/frontmatter.py
"""
Copyright 2026 Daniel Smith — Apache 2.0

YAML frontmatter extraction, merging, and serialization.

Three-stage pipeline the converter drives per file:

    text = path.read_text()
    fm, body = extract(text)
    merged = merge(fm, additions={"title": ..., "tags": [...], ...})
    out   = serialize(merged, body)
    path.write_text(out)

Merge rule of thumb: **user keys always win**. If a source file already
has `title: Something`, the converter's derived title is discarded. The
exception is `tags`, which is additive — existing and new tags are
unioned (deduped) so path-derived tags join, not replace, inline
frontmatter tags a user may have written.

Dependency: PyYAML. Install with `pip3 install --user pyyaml`,
`uv tool install pyyaml`, or a venv of your choice. See
`scripts/README.md` for options.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

try:
    import yaml
except ImportError as e:  # pragma: no cover
    raise ImportError(
        "pyyaml is required for the vault converter. "
        "Install with `pip3 install --user pyyaml` or via `uv`/venv. "
        "See scripts/README.md."
    ) from e


FENCE = "---"

# Matches an opening `---\n` or `---\r\n` at the very start of the file.
_OPEN_FENCE_RE = re.compile(r"\A---\r?\n")

# Matches a closing fence on its own line: `\n---\n`, `\n---\r\n`, or the
# same at end-of-string without a trailing newline. Multiline so ^/$ work
# per-line.
_CLOSE_FENCE_RE = re.compile(r"^---\s*$\r?\n?", re.MULTILINE)


@dataclass(frozen=True)
class Frontmatter:
    """
    Parsed YAML frontmatter block.

    `present=False` means the source had no frontmatter block (or a
    malformed one we refused to parse). `data` is always a dict — empty
    when absent — so callers don't need to None-check.
    """
    data: dict[str, Any] = field(default_factory=dict)
    present: bool = False


# --- extract ---------------------------------------------------------------


def extract(source: str) -> tuple[Frontmatter, str]:
    """
    Split `source` into `(Frontmatter, body)`.

    Recognises `---\\n…\\n---\\n` (and `\\r\\n` variants) at the start
    of the file. Malformed blocks — missing close fence, invalid YAML,
    non-dict top-level — return `Frontmatter(present=False)` and the
    full original source as body. No data loss.

    Ignores a leading UTF-8 BOM if present.
    """
    if source.startswith("\ufeff"):
        source = source[1:]

    open_match = _OPEN_FENCE_RE.match(source)
    if not open_match:
        return Frontmatter(), source

    after_open = source[open_match.end():]
    close_match = _CLOSE_FENCE_RE.search(after_open)
    if not close_match:
        # No close fence → treat as no frontmatter; don't accidentally
        # consume the whole file as YAML.
        return Frontmatter(), source

    yaml_text = after_open[:close_match.start()]
    body = after_open[close_match.end():]

    try:
        parsed = yaml.safe_load(yaml_text)
    except yaml.YAMLError:
        # Malformed YAML → treat as no frontmatter. Returning (empty,
        # original source) lets the converter still add its derived
        # fields without risking a data-loss overwrite of a file we
        # couldn't parse.
        return Frontmatter(), source

    if parsed is None:
        # Empty frontmatter block (`---\n---\n`) — present but no keys.
        return Frontmatter(data={}, present=True), body

    if not isinstance(parsed, dict):
        # A bare scalar or a list at the top level isn't real
        # frontmatter. Refuse to interpret, preserve the source.
        return Frontmatter(), source

    return Frontmatter(data=parsed, present=True), body


# --- merge -----------------------------------------------------------------


def merge(existing: Frontmatter, additions: dict[str, Any]) -> Frontmatter:
    """
    Merge `additions` into `existing.data` without clobbering user keys.
    Returns a fresh Frontmatter; does not mutate the input.

    Rules:
      - For any key in `additions` that is NOT already in `existing`,
        the addition's value is written.
      - `tags` is the one exception: new tags are unioned with existing
        tags (deduped), so path-derived tags join rather than replace
        whatever the user already declared.
      - Empty or None values in `additions` are skipped — we never write
        a meaningless `description: null` just because the source body
        was empty.
    """
    merged: dict[str, Any] = dict(existing.data)

    for key, value in additions.items():
        if key == "tags":
            union = _merge_tags(merged.get("tags"), value)
            if union:
                merged["tags"] = union
            continue

        if value is None:
            continue
        if isinstance(value, str) and not value.strip():
            continue

        if key not in merged:
            merged[key] = value
        # else: user key wins, preserve as-is.

    return Frontmatter(data=merged, present=True)


def _merge_tags(existing: Any, additions: Any) -> list[str]:
    """
    Union existing + addition tags, deduping by case-insensitive form
    with hyphens normalized to underscores. Output stores the
    underscore/lowercase form so Alap expressions can reference them
    (hyphens are the WITHOUT operator, so `.framework-guides` wouldn't
    parse as a single tag selector).
    """
    out: list[str] = []
    seen: set[str] = set()
    for source in (_as_tag_list(existing), _as_tag_list(additions)):
        for raw in source:
            normalized = raw.lstrip("#").strip().lower().replace("-", "_")
            if normalized and normalized not in seen:
                out.append(normalized)
                seen.add(normalized)
    return out


def _as_tag_list(raw: Any) -> list[str]:
    """
    Coerce a frontmatter `tags:` value into a list of strings. Handles:
      - None / missing → []
      - a bare string → [string]
      - a list of strings → as-is (non-string items stringified)
    Anything else (dict, nested list, …) is treated as empty — we
    don't pretend to understand exotic shapes.
    """
    if raw is None:
        return []
    if isinstance(raw, str):
        return [raw]
    if isinstance(raw, list):
        return [str(x) for x in raw if x is not None]
    return []


# --- serialize -------------------------------------------------------------


def serialize(fm: Frontmatter, body: str) -> str:
    """
    Emit the final markdown: frontmatter block (if any) + body.

    When `fm.data` is empty the block is omitted entirely — we don't
    emit `---\\n---\\n` just to say "there's no metadata." Body is
    passed through unchanged.

    PyYAML's `safe_dump` handles quoting, escaping, unicode. Keys stay
    in declaration order (relies on Python 3.7+ insertion-ordered
    dicts); `sort_keys=False` keeps the user's author-set keys from
    being reshuffled each regen.
    """
    if not fm.data:
        return body

    yaml_text = yaml.safe_dump(
        fm.data,
        default_flow_style=False,
        sort_keys=False,
        allow_unicode=True,
    )
    # safe_dump emits a trailing newline, so the close fence sits on its
    # own line without needing an explicit \n.
    return f"{FENCE}\n{yaml_text}{FENCE}\n{body}"
